AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestMatrixStore.kt

```diff
@@ -105,21 +105,89 @@ internal class TestMatrixStore(
+            clientInfo = clientInfo,
+            sharding = sharding,
+            deviceSetup = deviceSetup,
+            appApk = appApk,
+            testApk = testApk,
+            pullScreenshots = pullScreenshots,
+            testTargets = testTargets,
+            flakyTestAttempts = flakyTestAttempts
+        )
+        logger.info {
+            "created test matrix: $newTestMatrix"
+        }
+        // save it to cache, we don't worry about races here much such that if another instance happens to be creating
+        // the exact same test, one of them will win but that is OK since they'll each use their own test matrices and
+        // followup calls will re-use the winner of this race.
+        val testRun = TestRun(
+            id = testRunId,
+            testMatrixId = checkNotNull(newTestMatrix.testMatrixId) {
+                "newly created test matrix should not have null id $newTestMatrix"
+            }
+        )
+        datastoreApi.put(testRun.toEntity())
+        logger.info {
+            "saved test matrix info: $testRun"
+        }
+        return newTestMatrix
+    }
+
+    /**
+     * Creates a TestMatrix for the given configuration or returns an existing one if we've run the same tests
+     * specified in [testTargets] with the same environment configuration.
+     */
+    suspend fun getOrCreateTestMatrix(
+        testMatrix: TestMatrix,
+        cachedTestMatrixFilter: CachedTestMatrixFilter = { true },
+        testTargets: List<String>? = null,
+        flakyTestAttempts: Int = 2
+    ): TestMatrix {
+        checkNotNull(testMatrix.testMatrixId) {
+            "Test matrix id for the base test matrix should not be null"
+        }
+        val testRunId = TestRun.createId(
+            datastoreApi = datastoreApi,
+            environment = testMatrix.environmentMatrix,
+            clientInfo = testMatrix.clientInfo,
+            sharding = testMatrix.testSpecification.androidInstrumentationTest?.shardingOption,
+            testSetup = testMatrix.testSpecification.testSetup,
+            testTargets = testTargets,
+            testMatrixId = testMatrix.testMatrixId
+        )
+        logger.trace {
+            "test run id: $testRunId"
+        }
+
+        getExistingTestMatrix(testRunId)?.let {
+            logger.info("found existing test matrix: ${it.testMatrixId} with state: ${it.state}")
+            val state = it.state
+            // these states are ordered so anything above ERROR is not worth re-using
+            if (state != null && state >= TestMatrix.State.ERROR) {
+                logger.warn {
+                    "Skipping cache for ${it.testMatrixId} because its state is $state"
+                }
+            } else if (!cachedTestMatrixFilter(it)) {
+                logger.info {
+                    "Not re-using cached matrix due to filter"
+                }
+            } else {
+                return it
+            }
+        }
```

Comment:
this is duplicated from both methods, can you refactor it into a helper ?

Reply:
done